### PR TITLE
Sample "protocol_info" field

### PR DESCRIPTION
**Pull request by dongbohu**
## Issue Number
This PR addresses the backend for: https://github.com/AlexsLemonade/refinebio-frontend/issues/225

## Purpose/Implementation Notes
- Add a new `protocol_info` field to `Sample` model;
- Set `protocol_info` for each sample when a sample record is created by the GEO/ArrayExpress/SRA surveyor; 
- Expose `protocol_info` field in Sample API;

## Types of changes
- New feature (non-breaking change which adds functionality)

## Checklist

_Put an `x` in the boxes that apply._

- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: foreman/data_refinery_foreman/surveyor/sra.py [hunk 1]
```diff
@@ -422,6 +433,14 @@ def _generate_experiment_and_samples(self, run_accession: str) -> (Experiment, L
             for key, value in harmonized_sample.items():
                 setattr(sample_object, key, value)
 
+            protocol_info, is_updated = self.update_sample_protocol_info(
+                [],
```
Comment (Miserlou) [inline foreman/data_refinery_foreman/surveyor/sra.py]: Can you pass the kwarg name in here? It looks really weird to pass an empty array to as the first argument to a function

**Comment by Miserlou** [comment]
It looks like protocol_description is never set for GEO, is this intentional?

File: foreman/data_refinery_foreman/surveyor/sra.py (hunk 1 above)
Comment (dongbohu) [reply foreman/data_refinery_foreman/surveyor/sra.py]: Sure. I will.

**Comment by dongbohu** [comment]
@Miserlou GEO samples have sample-level protocol information, that's why GEO experiment's `protocol_description` is not set.

**Comment by dongbohu** [comment]
Here is the reason why the salmon tests failed:
https://github.com/AlexsLemonade/refinebio/blob/a9c51c2f67b72b014050572624afea280c6bc12d/workers/data_refinery_workers/processors/salmon.py#L222
because `OriginalFile` model doesn't have `get_synced_file_path` method.

@kurtwheeler and @Miserlou: Is this a change introduced by our collaborative code review?

**Comment by Miserlou** [comment]
Seems like it. I think that can be absolute_file_path instead?

**Comment by dongbohu** [comment]
@Miserlou Yes, `absolute_file_path` seems reasonable. Let me change it and see whether the tests will be happy.

**Comment by dongbohu** [comment]
@Miserlou and @kurtwheeler: Please let me know if you have any concerns about this PR. If not, I will close it.

File: foreman/data_refinery_foreman/surveyor/array_express.py [hunk 2]
```diff
@@ -270,6 +282,51 @@ def determine_sample_accession(self, experiment_accession: str, sample_source_na
...
+            for existing_protocol in existing_protocols:
+                if (new_protocol.get('accession', '') == existing_protocol['Accession']
+                    and new_protocol.get('text', '') == existing_protocol['Text']
+                    and new_protocol.get('type', '') == existing_protocol['Type']):
+                    new_protocol_is_found = True
+                    break
+            if not new_protocol_is_found:
+               existing_protocols.append({
+                   'Accession': new_protocol['accession'],
+                   'Text': new_protocol['text'],
```
Comment (Miserlou) [inline foreman/data_refinery_foreman/surveyor/array_express.py]: Suggest this all be moved to `.get()`

Comment (dongbohu) [reply foreman/data_refinery_foreman/surveyor/array_express.py]: I actually did it on purpose. Lines 308-309 guarantee that neither `accession` nor `text` field will be empty.

**Comment by dongbohu** [comment]
@Miserlou As you may have seen, I added a few test cases and all checks have passed. Please let me know whether you have any other comments. Thanks.

**Pull request merged**